specs/APIReview_CustomDownload.md

```diff
@@ -0,0 +1,346 @@
+
+    ```c#
+    void SampleMethod()
+    {
+        var show = new AnExampleOf();
+        show.SomeMembers = AndWhyItMight(be, interesting)
+    }
+    ```
+
+    ```cpp
+    void SampleClass::SampleMethod()
+    {
+        winrt::com_ptr<ICoreWebView2> webview2 = ...
+    }
+    ```
+
+    ## SecondFeatureName
+
+    Feature explanation text goes here, including why an app would use it, how it
+    replaces or supplements existing functionality.
+
+    ```c#
+    void SampleMethod()
+    {
+        var show = new AnExampleOf();
+        show.SomeMembers = AndWhyItMight(be, interesting)
+    }
+    ```
+
+    ```cpp
+    void SampleClass::SampleMethod()
+    {
+        winrt::com_ptr<ICoreWebView2> webview2 = ...
+    }
+    ```
+
+    As an example of this section, see the Examples section for the PasswordBox
+    control (https://docs.microsoft.com/windows/uwp/design/controls-and-patterns/password-box#examples).
+-->
+## Win32 C++
+```cpp
+// Register a handler for the `DownloadStarting` event.
+// This example hides the default downloads UI and shows a dialog box instead.
+// The dialog box displays the default save path and allows the user to specify a different path.
+// Selecting `OK` will save the download to the chosen path.
+// Selecting `CANCEL` will cancel the download.
+m_webViewStaging = m_webView.query<ICoreWebView2Staging2>();
+m_demoUri = L"https://demo.smartscreen.msft.net/";
+CHECK_FAILURE(m_webViewStaging->add_DownloadStarting(
+    Callback<ICoreWebView2StagingDownloadStartingEventHandler>(
+        [this](
+            ICoreWebView2Staging2* sender,
+            ICoreWebView2StagingDownloadStartingEventArgs* args) -> HRESULT {
+            // Hide the default downloads UI.
+            CHECK_FAILURE(args->put_HideUI(TRUE));
+
+            UINT64 totalBytes = 0;
+            CHECK_FAILURE(args->get_TotalBytes(&totalBytes));
+
+            wil::unique_cotaskmem_string uri;
+            CHECK_FAILURE(args->get_Uri(&uri));
+
+            wil::unique_cotaskmem_string mimeType;
+            CHECK_FAILURE(args->get_MimeType(&mimeType));
+
+            wil::unique_cotaskmem_string contentDisposition;
+            CHECK_FAILURE(args->get_ContentDisposition(&contentDisposition));
+
+            wil::unique_cotaskmem_string savePath;
+            CHECK_FAILURE(args->get_SavePath(&savePath));
+
+            std::wstring prompt =
+                std::wstring(L"Enter save path or select `OK` to use default path. ") +
+                L"Select `Cancel` to cancel the download.";
+
+            // Display the uri, mime type, and total bytes of the download.
+            std::wstring description = std::wstring(L"Uri: ") + uri.get() + L"\r\n" +
+                                        L"Mime type: " + mimeType.get() + L"\r\n" +
+                                        L"Total bytes: " + std::to_wstring(totalBytes) +
+                                        L"\r\n";
+
+            TextInputDialog dialog(
```

Comment:
This came up in [previous PRs](https://github.com/MicrosoftEdge/WebView2Feedback/pull/500#discussion_r499955638) that we don't want to have reentrancy during event handlers. So we don't want to show dialogs during the event handler. Changing this will also mean demonstrating how to use the deferral since we'll need to run code async later that shows the dialog and then completes the deferral.

Reply:
Sounds good, I updated the win32 sample to do this, but I still need to update the .net sample (it just uses the deferral in the event handler right now)